### ISSUE TITLE
Heart rate measurement features

### ISSUE THREAD

**Issue by cpfeiffer** [Freeyourgadget/Gadgetbridge#380]
This is an umbrella issue for several improvements and new features related to heart rate measurements:

#379: continuous heart rate measurements and recording
#377: Display heart rate in notification
#372: trigger alarm if heart rate is too high or too low

**Comment by rgh36167** [comment]
https://github.com/Freeyourgadget/Gadgetbridge/issues/381

**Comment by soder10** [comment]
Can you reply me with some Feedback regarding the following topic: can any 3rd parts app get the sleep tracking activity raw measurement values? I mean its the app developer who calculates the sleep intensity, or the firmware of the band already decides if I am in deep slow or not?  I want to see how much are my hearth rate measurements, not just if it is considered as deep or light? I hope it does not need live Bluetooth connection for it?

**Comment by rgh36167** [comment]
https://github.com/Freeyourgadget/Gadgetbridge/issues/382 : detect "strange" values and try to deal with them. In some cases a measurement can be repeated, in other cases it can be guessed that it should be half of the value or similar

**Comment by tomthe** [comment]
Request: Display resting-heart-rate for every night. I wrote a python script that displays a graph with the resting heart rate for every night and I used this SQL-code to get the value for one night from the Mi-Band 1s:

``` sql
select HEART_RATE, datetime(timestamp, 'unixepoch') as datum, timestamp
from MI_BAND_ACTIVITY_SAMPLE
where HEART_RATE between 27 and 160
and date(datetime(timestamp+3600*6, 'unixepoch'))=date('now', '-' || 25 || ' days')
order by HEART_RATE
LIMIT 1 OFFSET 15
```
it filters out unrealistic values and takes the 15th lowest value (to reduce noise). It would be awesome if Gadgetbridge would display this in the week-list

**Comment by rgh36167** [comment]
@tomthe: I think the approach could be slightly improved - many of the "unrealistic" values are half or multiple of the "real" pulse. This is probably in part due to measuring errors caused by hand movement and external light and in part because of pulse getting "blurry" under certain physiological conditions as well real extra systoles in some cases. A value of 60 followed by 120, 180 or 240 +-10% very likely is the result of one of these errors.

What could be done is to compare each value with a gliding window average of the last few values. If it is a multiple or fraction of the "last apparently valid frequency" look at the activity level to classify it as either
* likely error caused by movement
* possibly irregular heart beat
- and display the average of normal values as well as the percentages of the two known errors.

Btw for the first experiments a python script would be great, sqlite & python are easily usable for example with the termux package even on phones.

**Comment by tomthe** [comment]
@rgh36167 some of the unrealistic HR values really seem to be multiples of the real value. But not all. Interesting nevertheless and worth further investigation.
But for the calculation of the resting heart rate I think it doesn't matter.

**Comment by rgh36167** [comment]
@tomthe I will try to stratify the values against activity intentisty and number of steps as, probably using a few minutes sliding window average for activity and steps. On my device I am also getting enough unrealistic values that are too low that I can't simply take the minimum.

**Comment by tomthe** [comment]
I created a python-script that plots resting HR and steps per day:
You need to copy the Gadgetbridge sqlite database into the same directory.
https://gist.github.com/tomthe/dd50618cd4209aeafbc057d0bf01624c

**Comment by cpfeiffer** [comment]
Great job! It would be great if we could use this to somehow "calibrate" the measurements. I wouldn't be surprised if the adjustments would be slightly different for everybody.

**Comment by tomthe** [comment]
What do you want to calibrate? The HR-datapoints that are discarded (<20, >200)? The limit and offset (1, 20)?

**Comment by cpfeiffer** [comment]
Yeah, maybe even the window size. We'd need some data to find values that work for everybody, or do the calibration.
